### ISSUE TITLE
Cypress cannot write to the cache directory due to file permissions in Bitbucket pipeline

### ISSUE THREAD

**Issue by pieterjandesmedt** [cypress-io/cypress#15025]
### Current behavior

When running `npm install` in bitbucket pipelines, this happens:

```
> cypress@6.4.0 postinstall /opt/atlassian/pipelines/agent/build/node_modules/cypress
> node index.js --exec install
Cypress cannot write to the cache directory due to file permissions
See discussion and possible solutions at
https://github.com/cypress-io/cypress/issues/1281
----------
Failed to access /root/.cache/Cypress:
EACCES: permission denied, mkdir '/root/.cache/Cypress'
----------
```

### Desired behavior

Don't write to the root directory. This should be fixed since 4.6.0, according to #1281

### Test code to reproduce
Create a Bitbucket repository, enable pipelines, do `npm i` in the Bitbucket pipeline.

### Versions
6.4.0

**Comment by maxsuelmarinho** [comment]
Hi @pieterjandesmedt 

I solved this by adding `unset NPM_CONFIG_USER` into my `bitbucket-pipelines.yml` before the `npm ci` command. I found a similar issue [here](https://github.com/expo/sentry-expo/pull/26#issuecomment-453822980) and that solution worked for me


````yml
- step:
      name: Install dependencies
      caches:
        - npm
        - cypress
        - node
      script:
        - unset NPM_CONFIG_USER
        - npm ci

**Comment by n1ngu** [comment]
I reproduced this with a local node:alpine docker image, so this has nothing to do with Bitbucket Pipelines (but I found this issue there as well). I think it is more related to the default usage of the root user in docker containers, because sudoing into the "node" user fixes the issue.

**Comment by n1ngu** [comment]
@maxsuelmarinho that wasn't working for me, but this might be due to my usage of npm@7

I also tried exporting CYPRESS_CACHE_FOLDER=/tmp/foo-cypress-bar but then it simply fails due to lack of permissions on this other folder. But this is not a permission issue because
1. running the mkdir instructions in the shell does work (obviously)
2. I am root! I can do literally anything!

Honestly I don't see the point for being that picky with the cache. I'd expect it to be disabled if it didn't work, not to fail the whole npm installation.

Can the cache be somehow manually disabled?

**Comment by n1ngu** [comment]
Downgrading to npm@6 also solves the problem. Therefore this issue is a combination of using npm@7 and the root user.

**Comment by n1ngu** [comment]
For the record, this reproduces in Cypress version 7.7.0

**Comment by n1ngu** [comment]
Trying to narrow down the problem, running
```
$ docker run --rm --tty --interactive --workdir=/tmp/cypress --user=root node:14-alpine sh
# npm install --global npm@7
# npm install cypress
# ls -la ~/.cache/Cypress
```
worked just fine. Is this an issue in combination of some other dependencies? Now I am utterly astray.

**Comment by n1ngu** [comment]
Ok, the final trigger that puzzled me is the working directory being owned by a different user. I think this instruction
```
docker run --rm --workdir=/mnt/cypress --user=root node:14-alpine sh -c 'npm install --global npm@7 && chown --recursive node:node . && npm install cypress'
```
or
```
docker run --rm --workdir=/home/node --user=root node:14-alpine sh -c 'npm install --global npm@7 && npm install cypress'
```
are the minimal reproduction of the issue. Also reproduces with node-16 and non-alpine images.

Based on this, I'd hold to the hypothesis that permissions are improperly checked. Also opens the following workarounds:
1. Downgrade to npm@6
2. Sudo into the working dir owner
3. Chown the working dir to root

cc/ @pieterjandesmedt

**Comment by n1ngu** [comment]
Another workaround: depending on your workflow, exporting `CYPRESS_INSTALL_BINARY=0` will lead to a successful installation without the cypress binary. See https://docs.cypress.io/guides/getting-started/installing-cypress#Skipping-installation

You can later try to `npx cypress install` but it will fail because of the same issue.

**Comment by gustavodamazio** [comment]
Try using `yarn` package manager or downgrading `npm` to version 7.
For me it worked correctly.

I'm using **Jenkins** in an **alpine linux** container with **node 16**.

        stage('Prepare-build') {
        steps {
            sh 'node --version'
            sh 'npm --version'
            sh 'ls -p'
            sh 'git config --global --add safe.directory "*"'
            sh 'yarn install'
        }
    }

The error I was getting is the following:

    npm ERR! code 1
    npm ERR! path /home/jenkins/agent/workspace/project/app/project-web-manager/node_modules/cypress
    npm ERR! command failed
    npm ERR! command sh /tmp/postinstall-716e113e.sh
    npm ERR! Cypress cannot write to the cache directory due to file permissions
    npm ERR! 
    npm ERR! See discussion and possible solutions at
    npm ERR! https://github.com/cypress-io/cypress/issues/1281
    npm ERR! 
    npm ERR! ----------
    npm ERR! 
    npm ERR! Failed to access /root/.cache/Cypress:
    npm ERR! 
    npm ERR! EACCES: permission denied, mkdir '/root/.cache/Cypress'
    npm ERR! 
    npm ERR! ----------
    npm ERR! 
    npm ERR! Platform: linux-x64 (Alpine Linux - 3.16.1)
    npm ERR! Cypress Version: 10.6.0
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /root/.npm/_logs/2022-08-18T17_09_27_887Z-debug-0.log

In my same case export the environment variable cypress cache folder did not work as expected, the error was even made mentioned above.

    export CYPRESS_CACHE_FOLDER="/home/jenkins/agent/workspace/project/app/.cache"

**Comment by marcosaguilera-glober** [comment]
I was getting same issue:
**# EACCES: permission denied, mkdir '/root/.cache/Cypress'**

I've fixed like this:
1. In my Dockerfile
```
#Cypress version using
FROM cypress/included:10.6.0

# Installing sudo for superower installation required
# into entrypoint.sh
RUN apt-get install sudo

# Directories actions
WORKDIR /home/pwsh/tempVolume

ENTRYPOINT ["/bin/sh", "./entrypoint.sh"]
```

2. In my entrypoint.sh script file
```
#!/bin/sh
echo "I'm into the sh file"

## Install packages defined into package.json > dependencies node
##npm install
sudo npm install --unsafe-perm=true --allow-root

## Run cypress
cypress run
```
I messed up a time with this issue, and I tried with permissions `chown node:node`, also some `chmod -R 777`, but nothing worked for me, so I installed `sudo apt-get` into Linux distro.

Hope this help anyone.

**Comment by Rodeoclash** [comment]
I'm not sure if this will work around the problem for others, but seems to have for me:

```Dockerfile
FROM node:16-bullseye as development

EXPOSE 3000

ENV APP_HOME=/usr/src/app \
  CYPRESS_CACHE_FOLDER="/cypress/.cache" \
  LANG=C.UTF-8

WORKDIR $APP_HOME

RUN mkdir -p $CYPRESS_CACHE_FOLDER
COPY ./package.json ./
COPY ./package-lock.json ./

RUN npm install
```

The key here being that I made the cache folder first before running the `npm install`

YMMV!
